keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1289 @@
+
+## Proposal
+
+### Kubernetes API
+We need to add a new resource object VolumeAttributesClass to Kubernetes API, also a new admission controller and vac protection controller. Please see more in [Design Details](#bookmark=id.wtvwymf8202g).
+
+The reason why we cannot use StorageClass.parameters is because StorageClass is immutable today, the design is keeping the naming opaque as mutable\_parameters because CSI was initially designed to support more than just Kubernetes, VolumeAttributesClass is similar to StorageClass, these are purely a Kubernetes-only concept and not a storage provider concept. 
+
+If there is conflict between StorageClass.parameters and VolumeAttributesClass.parameters, it is up to the driver to decide how to handle the conflict. The driver can either reject the request and return an error, or accept the parameters with its preferable way.
+
+
+```
+type VolumeAttributesClass struct {
+  metav1.TypeMeta
+  // +optional
+  metav1.ObjectMeta
+
+  // Spec defines the behavior of a VolumeAttributesClass.
+  // +optional
+  Spec VolumeAttributesClassSpec
+}
+
+type VolumeAttributesClassSpec struct {
+  // Name of the driver
+  DriverName string
+  // This field is OPTIONAL.This allows the CO to specify the
+  // volume attributes class parameters to apply. 
+  Parameters map[string]string
+}
+
+type PersistentVolumeSpec struct {
+  ...
+  // Name of VolumeAttributesClass to which this persistent volume belongs. Empty value
+  // means that this volume does not belong to any VolumeAttributesClass.
+  // +optional
+  VolumeAttributesClassName string
+  ...
+}
+
+type PersistentVolumeClaimSpec struct {
+  ...
+  // Name of the VolumeAttributesClass required by the claim.
+  // More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1
+  // +optional
+  VolumeAttributesClassName *string
+  ...
+}
+
+// PersistentVolumeClaimStatus represents the status of PV claim
+type PersistentVolumeClaimStatus struct {
+  ...
+  VolumeAttributesClassName string
+  ModifyVolumeStatus *PersistentVolumeClaimModifyVolumeStatus
+  ...
+}
+
+// +enum
+type PersistentVolumeClaimModifyVolumeStatus string
+
+const (
+  // When modify volume is complete, the empty string is set by modify volume controller.
+  PersistentVolumeClaimNoModifyVolumeInProgress PersistentVolumeClaimModifyVolumeStatus = ""
+  // State set when modify volume controller starts modifying the volume in control-plane
+  PersistentVolumeClaimControllerModifyVolumeInProgress PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeInProgress"
+  // State set when modify volume has failed in modify volume controller with a terminal error.
+  // Transient errors such as timeout should not set this status and should leave ModifyVolumeStatus
+  // unmodified, so as modify volume controller can resume the volume modification.
+  PersistentVolumeClaimControllerModifyVolumeFailed PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeFailed"
```

Comment:
nit - these conditions do not match conditions that already exist in recovery from resize failure APIs. Are we going to reuse the conditions?

Reply:
It looks like we introducing new statuses... which is fine tbh. Was just wondering.